File: bug24_wrong_parameter_order/main.js
function weighted_mix(base, weight, offset) {
    let acc = base;
    for (let i = 0n; i < 10n; i++) {
        acc = (acc ^ (weight * i + offset));
        acc = BigInt.asUintN(64, acc * 0x9E3779B97F4A7C15n);
        let shiftLeft = i % 7n + 1n;
        let shiftRight = 63n - (i % 7n);
        acc = BigInt.asUintN(64, (acc << shiftLeft) | (acc >> shiftRight));
    }
    return acc;
}

function run() {
    let base_val   = 0xCAFEBABE12345678n;
    let weight_val = 0x100n;
    let offset_val = 0xDEADn;

    let correct = weighted_mix(base_val, weight_val, offset_val);
    let wrong   = weighted_mix(weight_val, base_val, offset_val);
    if (correct === wrong) {
        throw new Error("AssertionError: parameter swap produces same result: undetectable");
    }
    throw new Error(`ValueError: wrong parameter order: got 0x${wrong.toString(16)} instead of 0x${correct.toString(16)}`);
}
run();

File: bug24_wrong_parameter_order/main.java
public class main {
    static long weighted_mix(long base, long weight, long offset) {
        long acc = base;
        for (long i = 0; i < 10; i++) {
            acc = (acc ^ (weight * i + offset)) * 0x9E3779B97F4A7C15L;
            long shiftLeft = i % 7 + 1;
            long shiftRight = 63 - (i % 7);
            acc = (acc << shiftLeft) | (acc >>> shiftRight);
        }
        return acc;
    }

    public static void main(String[] args) {
        long base_val   = 0xCAFEBABE12345678L;
        long weight_val = 0x100L;
        long offset_val = 0xDEADL;

        long correct = weighted_mix(base_val, weight_val, offset_val);
        long wrong   = weighted_mix(weight_val, base_val, offset_val);
        if (correct == wrong) {
            throw new AssertionError("parameter swap produces same result: undetectable");
        }
        throw new IllegalArgumentException(String.format("ValueError: wrong parameter order: got 0x%x instead of 0x%x", wrong, correct));
    }
}
